File: admin/automated-tests/python-framework/src/services/lawyer_verification_service.py
import json
import time
import os
from typing import List, Dict, Any, Optional, Tuple
from dataclasses import dataclass
from difflib import SequenceMatcher
import re
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

@dataclass
class LawyerRecord:
    lastname: str
    firstname: str
    middle_name: str
    address: str
    roll_signed_date: str
    roll_no: str

@dataclass
class LawyerApplication:
    applicant_name: str
    first_name: str
    last_name: str
    middle_name: Optional[str] = None
    address: Optional[str] = None
    roll_number: Optional[str] = None
    application_id: str = ""

@dataclass
class VerificationResult:
    is_verified: bool
    confidence: float
    matches: List[LawyerRecord]
    match_type: str  # 'EXACT', 'PARTIAL', 'MULTIPLE', 'NONE'
    processing_time: float
    similarity_scores: List[float] = None

class LawyerVerificationService:
    def __init__(self, database_path: Optional[str] = None):
        self.database_path = database_path or os.path.join(
            os.path.dirname(__file__), '../../../data/lawyers_list.json'
        )
        self.lawyers_database: List[LawyerRecord] = []
        self.load_database()

    def load_database(self) -> None:
        """Load the lawyers database from JSON file."""
        try:
            start_time = time.time()
            with open(self.database_path, 'r', encoding='utf-8') as file:
                raw_data = json.load(file)
            
            self.lawyers_database = [
                LawyerRecord(
                    lastname=record.get('Lastname', ''),
                    firstname=record.get('Firstname', ''),
                    middle_name=record.get('Middle Name', ''),
                    address=record.get('Address', ''),
                    roll_signed_date=record.get('Roll Signed Date', ''),
                    roll_no=record.get('Roll No.', '')
                )
                for record in raw_data
            ]
            
            load_time = time.time() - start_time
            logger.info(f"Database loaded: {len(self.lawyers_database)} records in {load_time:.3f}s")
            
        except FileNotFoundError:
            logger.error(f"Database file not found: {self.database_path}")
            raise Exception("Database unavailable")
        except json.JSONDecodeError:
            logger.error(f"Invalid JSON in database file: {self.database_path}")
            raise Exception("Database unavailable")
        except Exception as e:
            logger.error(f"Failed to load database: {str(e)}")
            raise Exception("Database unavailable")

    def verify_lawyer(self, application: LawyerApplication) -> VerificationResult:
        """Verify a single lawyer application."""
        start_time = time.time()
        
        try:
            # Normalize application data
            normalized_app = self._normalize_application_data(application)
            
            # Search for matches
            matches, scores = self._find_matches(normalized_app)
            
            # Analyze matches
            result = self._analyze_matches(matches, scores, normalized_app)
            
            processing_time = time.time() - start_time
            
            return VerificationResult(
                is_verified=result['is_verified'],
                confidence=result['confidence'],
                matches=matches,
                match_type=result['match_type'],
                processing_time=processing_time,
                similarity_scores=scores
            )
            
        except Exception as e:
            logger.error(f"Verification failed: {str(e)}")
            return VerificationResult(
                is_verified=False,
                confidence=0.0,
                matches=[],
                match_type='NONE',
                processing_time=time.time() - start_time
            )

    def bulk_verify_lawyers(self, applications: List[LawyerApplication]) -> List[VerificationResult]:
        """Verify multiple lawyer applications."""
        start_time = time.time()
        results = []
        
        for i, application in enumerate(applications):
            logger.info(f"Processing application {i+1}/{len(applications)}: {application.applicant_name}")
            result = self.verify_lawyer(application)
            results.append(result)
        
        total_time = time.time() - start_time
        logger.info(f"Bulk verification completed: {len(applications)} applications in {total_time:.3f}s")
        
        return results

    def _normalize_application_data(self, application: LawyerApplication) -> LawyerApplication:
        """Normalize application data for consistent matching."""
        return LawyerApplication(
            applicant_name=self._normalize_string(application.applicant_name),
            first_name=self._normalize_string(application.first_name),
            last_name=self._normalize_string(application.last_name),
            middle_name=self._normalize_string(application.middle_name) if application.middle_name else None,
            address=self._normalize_string(application.address) if application.address else None,
            roll_number=application.roll_number,
            application_id=application.application_id
        )

    def _normalize_string(self, text: str) -> str:
        """Normalize string for consistent comparison."""
        if not text:
            return ""
        
        # Remove extra spaces and convert to uppercase
        normalized = re.sub(r'\s+', ' ', text.strip().upper())
        
        # Handle special characters commonly found in Filipino names
        normalized = normalized.replace('Ñ', 'N').replace('ñ', 'n')
        
        return normalized

    def _find_matches(self, application: LawyerApplication) -> Tuple[List[LawyerRecord], List[float]]:
        """Find matching lawyers in the database."""
        matches = []
        scores = []
        
        for lawyer in self.lawyers_database:
            score = self._calculate_similarity_score(application, lawyer)
            if score > 0.6:  # 60% similarity threshold
                matches.append(lawyer)
                scores.append(score)
        
        # Sort by similarity score (highest first)
        if matches:
            sorted_pairs = sorted(zip(matches, scores), key=lambda x: x[1], reverse=True)
            matches, scores = zip(*sorted_pairs)
            matches, scores = list(matches), list(scores)
        
        return matches, scores

    def _calculate_similarity_score(self, application: LawyerApplication, lawyer: LawyerRecord) -> float:
        """Calculate similarity score between application and lawyer record."""
        total_score = 0.0
        total_weight = 0.0
        
        # First name comparison (weight: 30%)
        if application.first_name and lawyer.firstname:
            first_name_score = self._string_similarity(
                application.first_name, 
                self._normalize_string(lawyer.firstname)
            )
            total_score += first_name_score * 0.3
            total_weight += 0.3
        
        # Last name comparison (weight: 40%)
        if application.last_name and lawyer.lastname:
            last_name_score = self._string_similarity(
                application.last_name, 
                self._normalize_string(lawyer.lastname)
            )
            total_score += last_name_score * 0.4
            total_weight += 0.4
        
        # Middle name comparison (weight: 20%)
        if application.middle_name and lawyer.middle_name:
            middle_name_score = self._string_similarity(
                application.middle_name, 
                self._normalize_string(lawyer.middle_name)
            )
            total_score += middle_name_score * 0.2
            total_weight += 0.2
        
        # Address comparison (weight: 10%)
        if application.address and lawyer.address:
            address_score = self._string_similarity(
                application.address, 
                self._normalize_string(lawyer.address)
            )
            total_score += address_score * 0.1
            total_weight += 0.1
        
        return total_score / total_weight if total_weight > 0 else 0.0

    def _string_similarity(self, str1: str, str2: str) -> float:
        """Calculate string similarity using SequenceMatcher."""
        if not str1 or not str2:
            return 0.0
        
        if str1 == str2:
            return 1.0
        
        return SequenceMatcher(None, str1, str2).ratio()

    def _analyze_matches(self, matches: List[LawyerRecord], scores: List[float], 
                        application: LawyerApplication) -> Dict[str, Any]:
        """Analyze matches and determine verification result."""
        if not matches:
            return {
                'is_verified': False,
                'confidence': 0.0,
                'match_type': 'NONE'
            }
        
        if len(matches) == 1:
            confidence = scores[0] * 100
            return {
                'is_verified': confidence > 90,
                'confidence': round(confidence, 1),
                'match_type': 'EXACT' if confidence > 90 else 'PARTIAL'
            }
        
        # Multiple matches
        best_score = scores[0] * 100
        return {
            'is_verified': best_score > 95,  # Higher threshold for multiple matches
            'confidence': round(best_score, 1),
            'match_type': 'MULTIPLE'
        }

    def get_database_stats(self) -> Dict[str, Any]:
        """Get database statistics."""
        try:
            file_size = os.path.getsize(self.database_path)
            return {
                'total_records': len(self.lawyers_database),
                'database_size': f"{file_size / 1024 / 1024:.2f} MB",
                'file_path': self.database_path
            }
        except Exception as e:
            logger.error(f"Failed to get database stats: {str(e)}")
            return {
                'total_records': len(self.lawyers_database),
                'database_size': 'Unknown',
                'file_path': self.database_path
            }

    def search_lawyers_by_name(self, name: str, limit: int = 10) -> List[LawyerRecord]:
        """Search lawyers by name (utility method for testing)."""
        normalized_name = self._normalize_string(name)
        matches = []
        
        for lawyer in self.lawyers_database:
            full_name = f"{lawyer.firstname} {lawyer.lastname}".strip()
            if normalized_name.lower() in self._normalize_string(full_name).lower():
                matches.append(lawyer)
                if len(matches) >= limit:
                    break
        
        return matches

File: admin/automated-tests/js-framework/src/services/LawyerVerificationService.ts
import fs from 'fs';
import path from 'path';
import _ from 'lodash';

export interface LawyerRecord {
  Lastname: string;
  Firstname: string;
  'Middle Name': string;
  Address: string;
  'Roll Signed Date': string;
  'Roll No.': string;
}

export interface LawyerApplication {
  applicant_name: string;
  first_name: string;
  last_name: string;
  middle_name?: string;
  address?: string;
  roll_number?: string;
  application_id: string;
}

export interface VerificationResult {
  isVerified: boolean;
  confidence: number;
  matches: LawyerRecord[];
  matchType: 'EXACT' | 'PARTIAL' | 'MULTIPLE' | 'NONE';
  processingTime: number;
}

export class LawyerVerificationService {
  private lawyersDatabase: LawyerRecord[] = [];
  private databasePath: string;

  constructor(databasePath?: string) {
    this.databasePath = databasePath || path.join(__dirname, '../../../data/lawyers_list.json');
    this.loadDatabase();
  }

  private loadDatabase(): void {
    try {
      const startTime = Date.now();
      const rawData = fs.readFileSync(this.databasePath, 'utf8');
      this.lawyersDatabase = JSON.parse(rawData);
      const loadTime = Date.now() - startTime;
      console.log(`Database loaded: ${this.lawyersDatabase.length} records in ${loadTime}ms`);
    } catch (error) {
      console.error('Failed to load lawyers database:', error);
      throw new Error('Database unavailable');
    }
  }

  public async verifyLawyer(application: LawyerApplication): Promise<VerificationResult> {
    const startTime = Date.now();
    
    try {
      // Normalize application data
      const normalizedApp = this.normalizeApplicationData(application);
      
      // Search for matches
      const matches = this.findMatches(normalizedApp);
      
      // Determine verification result
      const result = this.analyzeMatches(matches, normalizedApp);
      
      const processingTime = Date.now() - startTime;
      
      return {
        ...result,
        processingTime
      };
    } catch (error) {
      return {
        isVerified: false,
        confidence: 0,
        matches: [],
        matchType: 'NONE',
        processingTime: Date.now() - startTime
      };
    }
  }

  public async bulkVerifyLawyers(applications: LawyerApplication[]): Promise<VerificationResult[]> {
    const startTime = Date.now();
    const results: VerificationResult[] = [];
    
    for (const application of applications) {
      const result = await this.verifyLawyer(application);
      results.push(result);
    }
    
    const totalTime = Date.now() - startTime;
    console.log(`Bulk verification completed: ${applications.length} applications in ${totalTime}ms`);
    
    return results;
  }

  private normalizeApplicationData(application: LawyerApplication): LawyerApplication {
    return {
      ...application,
      first_name: application.first_name?.trim().toUpperCase() || '',
      last_name: application.last_name?.trim().toUpperCase() || '',
      middle_name: application.middle_name?.trim().toUpperCase() || '',
      address: application.address?.trim() || ''
    };
  }

  private findMatches(application: LawyerApplication): LawyerRecord[] {
    const matches: LawyerRecord[] = [];
    
    for (const lawyer of this.lawyersDatabase) {
      const score = this.calculateSimilarityScore(application, lawyer);
      if (score > 0.6) { // 60% similarity threshold
        matches.push(lawyer);
      }
    }
    
    return matches.sort((a, b) => 
      this.calculateSimilarityScore(application, b) - 
      this.calculateSimilarityScore(application, a)
    );
  }

  private calculateSimilarityScore(application: LawyerApplication, lawyer: LawyerRecord): number {
    let score = 0;
    let totalWeight = 0;
    
    // First name comparison (weight: 30%)
    const firstNameScore = this.stringSimilarity(
      application.first_name, 
      lawyer.Firstname.toUpperCase()
    );
    score += firstNameScore * 0.3;
    totalWeight += 0.3;
    
    // Last name comparison (weight: 40%)
    const lastNameScore = this.stringSimilarity(
      application.last_name, 
      lawyer.Lastname.toUpperCase()
    );
    score += lastNameScore * 0.4;
    totalWeight += 0.4;
    
    // Middle name comparison (weight: 20%)
    if (application.middle_name && lawyer['Middle Name']) {
      const middleNameScore = this.stringSimilarity(
        application.middle_name, 
        lawyer['Middle Name'].toUpperCase()
      );
      score += middleNameScore * 0.2;
      totalWeight += 0.2;
    }
    
    // Address comparison (weight: 10%)
    if (application.address && lawyer.Address) {
      const addressScore = this.stringSimilarity(
        application.address.toUpperCase(), 
        lawyer.Address.toUpperCase()
      );
      score += addressScore * 0.1;
      totalWeight += 0.1;
    }
    
    return totalWeight > 0 ? score / totalWeight : 0;
  }

  private stringSimilarity(str1: string, str2: string): number {
    if (str1 === str2) return 1;
    if (!str1 || !str2) return 0;
    
    // Simple Levenshtein distance-based similarity
    const longer = str1.length > str2.length ? str1 : str2;
    const shorter = str1.length > str2.length ? str2 : str1;
    
    if (longer.length === 0) return 1;
    
    const distance = this.levenshteinDistance(longer, shorter);
    return (longer.length - distance) / longer.length;
  }

  private levenshteinDistance(str1: string, str2: string): number {
    const matrix = Array(str2.length + 1).fill(null).map(() => Array(str1.length + 1).fill(null));
    
    for (let i = 0; i <= str1.length; i++) matrix[0][i] = i;
    for (let j = 0; j <= str2.length; j++) matrix[j][0] = j;
    
    for (let j = 1; j <= str2.length; j++) {
      for (let i = 1; i <= str1.length; i++) {
        const indicator = str1[i - 1] === str2[j - 1] ? 0 : 1;
        matrix[j][i] = Math.min(
          matrix[j][i - 1] + 1,
          matrix[j - 1][i] + 1,
          matrix[j - 1][i - 1] + indicator
        );
      }
    }
    
    return matrix[str2.length][str1.length];
  }

  private analyzeMatches(matches: LawyerRecord[], application: LawyerApplication): Omit<VerificationResult, 'processingTime'> {
    if (matches.length === 0) {
      return {
        isVerified: false,
        confidence: 0,
        matches: [],
        matchType: 'NONE'
      };
    }
    
    if (matches.length === 1) {
      const confidence = this.calculateSimilarityScore(application, matches[0]);
      return {
        isVerified: confidence > 0.9,
        confidence: Math.round(confidence * 100),
        matches,
        matchType: confidence > 0.9 ? 'EXACT' : 'PARTIAL'
      };
    }
    
    // Multiple matches
    const bestMatch = matches[0];
    const confidence = this.calculateSimilarityScore(application, bestMatch);
    
    return {
      isVerified: confidence > 0.95, // Higher threshold for multiple matches
      confidence: Math.round(confidence * 100),
      matches,
      matchType: 'MULTIPLE'
    };
  }

  public getDatabaseStats(): { totalRecords: number; databaseSize: string } {
    const stats = fs.statSync(this.databasePath);
    return {
      totalRecords: this.lawyersDatabase.length,
      databaseSize: `${(stats.size / 1024 / 1024).toFixed(2)} MB`
    };
  }
}
